### PR TITLE
Relax app dependencies for `lightning` | breaking imports from `L`

### PR DESCRIPTION
**Pull request by Borda**
## What does this PR do?

Alternative to #15653 (which would suffer from missing typing)
so if you install `lightning_app`, nothing changes
but if you install `lightning` now you would need to install `lightning[app]`

<details>
  <summary><b>Before submitting</b></summary>

- Was this **discussed/agreed** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- Did you make sure to **update the documentation** with your changes? (if necessary)
- Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- Did you list all the **breaking changes** introduced by this pull request?
- Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

</details>

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

<details>
  <summary>Reviewer checklist</summary>

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

</details>

cc @borda @carmocca @justusschock @awaelchli @tchaton

### CODE REVIEW

File: .github/workflows/ci-pkg-install.yml [hunk 1]
```diff
@@ -42,10 +42,13 @@ jobs:
...
       fail-fast: false
       matrix:
-        os: [ubuntu-22.04, macOS-12, windows-2022]
+        os: ["ubuntu-22.04", "macOS-12", "windows-2022"]
         pkg-name: ["app", "fabric", "pytorch", "lightning", "notset"]
         python-version: ["3.8" , "3.10"]
         # TODO: add also install from source
+        include:
+          - {os: "macOS-12", pkg-name: "lightning", python-version: "3.9", pkg-extra: "app"}
+          - {os: "ubuntu-22.04", pkg-name: "lightning", python-version: "3.9", pkg-extra: "app"}
```
Comment (Borda) [inline .github/workflows/ci-pkg-install.yml:51]:
```suggestion
          - {os: "macOS-12", pkg-name: "lightning", python-version: "3.9", pkg-extra: "app"}
          - {os: "macOS-12", pkg-name: "notset", python-version: "3.9", pkg-extra: "app"}
          - {os: "ubuntu-22.04", pkg-name: "lightning", python-version: "3.9", pkg-extra: "app"}
          - {os: "ubuntu-22.04", pkg-name: "notset", python-version: "3.9", pkg-extra: "app"}
```

**Comment by Borda** [comment]
the docs is fixed, see https://github.com/Lightning-AI/lightning/actions/runs/5973075464/job/16204667924?pr=18386

**Review by lantiga — APPROVED** [review]
Looks good, the question is what to do for `lightning.data`.

`LightningIterableDataset` doesn't depend on any cloud features.
`LightningDataset` only relies on `torchdata` and `botocore`. 

It would be enough to relax this dependency:

https://github.com/Lightning-AI/lightning/blob/master/src/lightning/data/datasets/index.py#L91

or provide coverage for the case where data connections are not available (which amounts to falling back to getting the index file directly through boto)

If we do these two we could keep `data` in `lightning`, which would be good IMO.

**Comment by Borda** [comment]
> the question is what to do for `lightning.data`.

sure, lets address it in a followup PR

File: src/lightning/__init__.py [hunk 2]
```diff
@@ -14,36 +16,38 @@
...
     "LightningDataModule",
     "LightningModule",
     "Callback",
     "seed_everything",
     "Fabric",
-    "storage",
 ]
+
+# todo: try to parse if the foiling is because of dependency or just wrong import
+_ABLE_LOAD_APP = module_available("lightning.app")
```
Comment (awaelchli) [inline src/lightning/__init__.py:35]:
```suggestion
_CAN_LOAD_APP = module_available("lightning.app")
```

File: src/lightning/__setup__.py [hunk 3]
```diff
@@ -39,20 +39,32 @@ def _prepare_extras() -> Dict[str, Any]:
...
     extras["data-dev"] = extras["data-all"] + extras["data-test"]
+    extras["store-store"] = extras["app-app"]  # todo: consider cutting/leaning this dependency
+
     # merge per-project extras of the same category, e.g. `app-test` + `fabric-test`
     for extra in list(extras):
         name = "-".join(extra.split("-")[1:])
         extras[name] = extras.get(name, []) + extras[extra]
+
+    # drop quasi base the req. file has the same name sub-package
+    for k in list(extras.keys()):
```
Comment (awaelchli) [inline src/lightning/__setup__.py:62]:
```suggestion
    for extra in extras:
```

**Review by awaelchli — APPROVED** [review]
I tested out this branch by running

```
pip install git+https://github.com/lightning-ai/lightning@refactor/app-root
```
in a clean environment and as expected only the pytorch related dependencies were installed. Great work!

However note, importing app now results in
```
>>> import lightning.app as app
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/adrian/miniconda3/envs/test-app-dep/lib/python3.10/site-packages/lightning/app/__init__.py", line 24, in <module>
    from lightning.app import components  # noqa: E402, F401
  File "/Users/adrian/miniconda3/envs/test-app-dep/lib/python3.10/site-packages/lightning/app/components/__init__.py", line 1, in <module>
    from lightning.app.components.database.client import DatabaseClient
  File "/Users/adrian/miniconda3/envs/test-app-dep/lib/python3.10/site-packages/lightning/app/components/database/__init__.py", line 1, in <module>
    from lightning.app.components.database.client import DatabaseClient
  File "/Users/adrian/miniconda3/envs/test-app-dep/lib/python3.10/site-packages/lightning/app/components/database/client.py", line 22, in <module>
    from lightning.app.components.database.utilities import _GeneralModel
  File "/Users/adrian/miniconda3/envs/test-app-dep/lib/python3.10/site-packages/lightning/app/components/database/utilities.py", line 20, in <module>
    from fastapi import Response, status
ModuleNotFoundError: No module named 'fastapi'

```

We would need to find a way to control the error message here, instructing the user to run the right install command for the extra.

File: src/lightning/__setup__.py (hunk 3 above)
Comment (Borda) [reply src/lightning/__setup__.py:62]: not really, because in this case you can't edit the dst you are iterating

File: src/lightning/__init__.py [hunk 4]
```diff
@@ -14,36 +16,38 @@
...
+_ABLE_LOAD_DATA = module_available("lightning.data")
```
Comment (Borda) [inline src/lightning/__init__.py:36]:
```suggestion
_CAN_LOAD_DATA = module_available("lightning.data")
```

File: src/lightning/__init__.py [hunk 5]
```diff
@@ -14,36 +16,38 @@
...
+
+
+if _ABLE_LOAD_APP:
```
Comment (Borda) [inline src/lightning/__init__.py:39]:
```suggestion
if _CAN_LOAD_APP:
```

Comment (Borda) [inline src/lightning/__init__.py:39]:
```suggestion
if module_available("lightning.app"):
```

File: src/lightning/__init__.py [hunk 6]
```diff
@@ -14,36 +16,38 @@
...
+    from lightning.app.core.app import LightningApp
+    from lightning.app.core.flow import LightningFlow
+    from lightning.app.core.work import LightningWork
+    from lightning.app.utilities.packaging.build_config import BuildConfig
+    from lightning.app.utilities.packaging.cloud_compute import CloudCompute
+
+    __all__ += ["LightningApp", "LightningFlow", "LightningWork", "BuildConfig", "CloudCompute", "storage"]
+
+
+if _ABLE_LOAD_DATA:
```
Comment (Borda) [inline src/lightning/__init__.py:50]:
```suggestion
if module_available("lightning.data"):
```

File: src/lightning/__init__.py (hunk 4 above)
Comment (Borda) [inline src/lightning/__init__.py:36]:
```suggestion
```

File: src/lightning/__init__.py [hunk 7]
```diff
@@ -12,38 +15,46 @@
...
+
+if module_available("lightning.data"):
+    from lightning.data import LightningDataset, LightningIterableDataset
+
+    __all__ += ["LightningDataset", "LightningIterableDataset"]
+else:
+    warnings.warn(
+        "Importing Data's sub-pckage failed."
+        " If you intend to use them, please install required dependencies `pip install lighting[data]`."
+    )
```
Comment (Borda) [inline src/lightning/__init__.py:60]:
```suggestion
```
fliping to breaking chnage...
cc: @lantiga

**Comment by Borda** [comment]
Note that the difference grew from 100 lines to 600 lines because we decided to go with breaking change instead of soft change, ref: https://github.com/Lightning-AI/lightning/pull/18386#discussion_r1305809775

**Pull request merged**